Add `SkeletonModifier3D` IKs as `IKModifier3D`

TokageItLab opened the pull request:
~Add several IK modifiers. As the name `SkeletonIK3D` conflicts with an older node, I will adopt the provisional name `ManyBoneIK3D`. Considering that the LookAtModifier targets only a single bone, I think this is a reasonable name.~ Tentatively it now named as `IKModifier3D`.

<img width="218" height="180" alt="image" src="https://github.com/user-attachments/assets/dcf0994f-bc48-4a93-80b2-4e50d0e925bf" />

ManyBoneIK3D has the following extended classes:

- IKModifier3D
  - TwoBoneIK3D
  - ChainIK3D
    - SplineIK3D
    - IterateIK3D
      - FABRIK3D
      - CCDIK3D
      - JacobianIK3D

There may be demand for FullbodyIK3D, but for now, this PR aims to fulfill the need for SkeletonModificationStack3D which was removed from 3.x to 4.0. (So don't close - https://github.com/godotengine/godot-proposals/issues/6039)

I expect that FullbodyIK3D could be implemented in the future by extending one of the base classes. Until then, I assume PhysicalBoneSimulator can serve as a substitute in some extent.

Also this PR includes refactoring to unify enums for bone direction among some SkeletonModifiers.

## IKModifier3D

It has a virtual method for having multiple configurations and a virtual bone structure for simulating IK to be able to hold poses and rests separately from the Skeleton.

## TwoBoneIK3D

It provides deterministic results by constructing a plane from each joint and pole target and finding the intersection of two circles (disks in 3D). This IK can handle twist by setting the knuckle direction.

<img width="374" height="269" alt="image" src="https://github.com/user-attachments/assets/a21cc76e-aa69-4824-a6fc-dfd14c272252" />

https://github.com/user-attachments/assets/0439c2d3-877e-4057-9ea5-1ef4dde4e0ab

If there are more than one bones between each set bone, their rotations are ignored, and the straight line connecting the root-middle and middle-end joints are treated as virtual bones.

https://github.com/user-attachments/assets/957bc4e6-b97c-45db-9d81-2ee8eec37825

## ChainIK3D

It has a virtual method that generates a chain by setting root bone and end bone.

## SplineIK3D

This IK aligning bones on Path3D. Bone twist is determined relatively based on the curve's tilt. 

https://github.com/user-attachments/assets/150cc831-7155-458c-9206-024d814bd703

If the root bone joint and the start point of Curve3D are apart, it assumes that there is a linear line segment between them. If the end bone joint exceeds the path length, bend the first joint exceeding the path length as close as possible to the end point of the Curve3D, and from there it is extended straight ahead - it should look almost the same as Blender's SplineIK.

## IterateIK3D

It has a virtual method to approach the goal by repeating small rotations.

- FABRIK3D
  - forward and backward reaching
  - good result when there are no limitations
- CCDIK3D
  - cyclic coordinate descent
  - good result for mechanical movements
- JacobianIK3D
  - Jacobian transpose
  - good result for biological movements

https://github.com/user-attachments/assets/408ef731-c988-48e1-81c0-cfbb98ec8813

Each bone chain (setting) has one effector, which is processed in order of setting list. You can set some limitations for each joint.

<img width="131" height="251" alt="image" src="https://github.com/user-attachments/assets/b35fcdae-b27b-45b3-803b-b68b747be23f" />

Limitations currently only support rotation axis and cone shapes. You can set arbitrary rotation offsets on each joint. BTW, there is potential for future reuse in SpringBone (ref. https://github.com/vrm-c/vrm-specification/pull/496). Also I expect that kusudama shapes like https://github.com/godotengine/godot-proposals/issues/11208 will be added in the future.

**Known issue**
Moving the position breaks the rendering of the Limitation gizmo because binding to position alone is impossible. To fix it, an approach such as canceling rotation within the shader like https://github.com/godotengine/godot-proposals/issues/9735 is required. Currently, it is drawn by a hack to bind the limitation to the *parent* joint. See also https://github.com/godotengine/godot/pull/111815.

### Incompatibility with old SkeletonIK3D

Regarding the pole target (magnet), I tried implementing it but finally decided against it because I couldn't stabilize it. This was due to strong conflicts with limitations and the difficulty in acquiring good results with solvers other than FABRIK. Even with FABRIK (both the version implemented this time and the current SkeletonIK3D), it is difficult to consistently obtain ideal results.

For these purposes, I determined it is more appropriate to use IK that can deterministically control direction, such as the TwoBoneIK and SplineIK described above, depending on the use case.

If someone needs a magnet no matter what, it possibly could be added later as a follow up PR. (If it can be stabilized)

If you want rotation on a plane, you can set the RotationAxis for absolute axis specification. For dynamic planes, you can use a LookAtModifier to orient the chain root specific (knuckle) axis to secondary (or same) target before IterateIK, which should achieve that. In other words, probably the general (other software's) chain IK's pole is processed in this way - only determining the twist of the first joint.

Also, omitted the functionality to apply the target's rotation to the tip. If we needed, the correct approach should be to allow users to specify a node in BoneConstraint3D (CopyTransformModifier3D) for arbitrary rotation transformations, rather than hard-coding it into the IK. See also https://github.com/godotengine/godot/pull/110336.

---

### Demo project

[ik_test.zip](https://github.com/user-attachments/files/22058952/ik_test.zip)

realcoloride commented:
Cool PR

TokageItLab committed:
Add IKModifier3D

lyuma commented:
We reviewed and tested this in the animation meeting

- Knuckle naming should be changed “pole direction”?
- Kusudama visualization - separate pr
- FIRE - Submit to godot demo projects
- Fix / Verify modifier animation baker: Wanted!
- Saracen described feature enhancement - Convert FK to IK animations.
- bug: CCDIK/JacobianIK target origin calculator is wrong: needs to fixed

We noticed the array properties do not have documentation. Since IK is very complex and there are a lot of arrays, adding documentation for nested properties is going to be increasingly important. We need to open a discussion with the documentation team. (This is also an issue for skeleton options in the import retargeter system). We need to open a dedicate tracker issue for this problem.

Overall, we are very happy with the design. If we fix some of the above I think we can approve this once we review the code.

TokageItLab commented:
- Renamed KnuckleDirection to PoleDirection
- Fixed a broken origin calculation
	- The cause was the root bone of the chain incorrectly obtaining its parent's origin. In this case, IK starting from the origin remained intact, so we had overlooked it..

wagnerfs commented:
I'm not sure if I can just butt myself in but I'm highly excited for this PR.
I'm having trouble with SplineIK3D, am I doing something wrong? (Therere 5 bones on the chain)


https://github.com/user-attachments/assets/9693313a-e745-406c-be7a-1bc82d27c31a

TokageItLab commented:
@wagnerfs The bake interval looks to be too large. Also, verify that the start and end points are not inverted; The bone's ancestor should be closer to the start of the curve.

ettiSurreal commented:
Tried using this for procedural animation. From my testing it seems to work really well, though I will share a few thoughts/suggestions:

- It would be useful to have a button on the top bar that would reset the IK, since currently to do that you have to toggle the modifier active on and off.
- IterateIK3D could have a deterministic toggle, that would work by always resetting the IK before iterating, which should make the bones always end up in the same pose based on the target placement, at the cost of needing more iterations therefore being more performance intensive, though I tried setting it to an insanely high iteration count and it seemed to perform fine, so it might be fine in that regard.
(I'm also pretty sure this is how IK in Blender operates, I think it uses the Jacobian algorithm, and always resets the IK before iterating it)
- There should be a way to lock a specific axis from rotation. Currently it's only possible to constraint rotation to only one axis but it would be nice if it were possible to also constraint it to _all but one_ axes. It would come in handy in cases like this one, where I want the IK to work on all axes but Y/Roll.
<details>
<summary>Video Example</summary>

https://github.com/user-attachments/assets/0f29c3d3-7e6e-4973-a538-76c6f2c5a869

</details>

- On a related note, it might also be useful to have a "None" rotation axis mode, which would completely disable the rotation of that specific bone. I think this can already be done by using JointLimitationCone3D with 0 radius, but this might be more convenient.
- Blender gives you an option to influence an IK chain by the target's rotation, which gives you an extra degree of control like such:
If this is deemed useful it would definitely be material for a follow up PR though.

<details>
<summary>Video Example</summary>

https://github.com/user-attachments/assets/37d1d5be-e998-4991-84f9-fa039ce02e7f

</details>

- Another one related to rotation, it would also be very useful to have an option to set the end bone rotation to the rotation of the target, for example so you can keep a foot aligned to the ground.
This could also work as it's own modifier, and during my testing I made a simple one for that purpose, which I will share in case anyone finds it useful:
<details>
<summary>CopyTargetRotationModifier3D</summary>
A skeleton modifier that aligns the bone based on the target node in world space, while completely cancelling out any external transformation.

```py
@tool
class_name CopyTargetRotationModifier3D
extends SkeletonModifier3D

@export var target: Node3D
@export var bone_index: int

func _process_modification_with_delta(_delta: float) -> void:
	get_skeleton().set_bone_global_pose(bone_index, Transform3D(target.global_basis * get_skeleton().global_basis.inverse(), get_skeleton().get_bone_global_pose(bone_index).origin))
```

</details>

TokageItLab commented:
@ettiSurreal Hi, thank you for the feedback!

> It would be useful to have a button on the top bar that would reset the IK, since currently to do that you have to toggle the modifier active on and off.

I'm not sure whether we should escalate to add a reset function to SkeletonModifier3D's BaseClass (since an implementation that simply toggles active might suffice), but I think we can consider it in a later PR.

However, I believe there are two scenarios: resetting a single modifier versus resetting multiple modifiers. When a modifier is selected, it should reset that single modifier; when a skeleton is selected, it should reset all modifiers.

---

> There should be a way to lock a specific axis from rotation. Currently it's only possible to constraint rotation to only one axis but it would be nice if it were possible to also constraint it to all but one axes. It would come in handy in cases like this one, where I want the IK to work on all axes but Y/Roll.

I don't understand this well. Regarding IterateIK, since it only handles Swing, the Twist axis (parent-to-child between chains) is always locked. The RotationAxis option selects and locks the second axis projected onto the plane normal to the Twist axis. Custom mode allows you to specify any second axis.

<img width="458" height="196" alt="image" src="https://github.com/user-attachments/assets/a926fd3e-ca4c-47d5-ba4c-74cf98bfb921" />

If you're saying you want to perform Twist with IterateIK, that might require a different algorithm than FABR/CCD/Jacobian.

---

> CopyTargetRotationModifier3D
> A skeleton modifier that aligns the bone based on the target node in world space, while completely cancelling out any external transformation.

This has been sent in https://github.com/godotengine/godot/pull/110336 already.

Regarding ground matching, it would likely involve combining with ray tracing or similar techniques. This would require a setup where the target moves outside the IK system, which should probably be provided as an add-on and falls outside the scope of this PR.

Options for constructing specific scenes, like 2D/3D platformer templates or creating PhysicalBones, could be implemented in additional PRs.

<img width="259" height="218" alt="image" src="https://github.com/user-attachments/assets/d6d89742-99ef-40a2-b48b-f58530db7f67" />

yosimba2000 commented:
Hi, does this new IK system adjust the position of the colliders attached to the bones as well?

AThousandShips commented on editor/scene/3d/gizmos/two_bone_ik_3d_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,207 @@
+
+shader_type spatial;
+render_mode unshaded, shadows_disabled;
+void vertex() {
```
```suggestion

void vertex() {
```

AThousandShips commented on editor/scene/3d/gizmos/chain_ik_3d_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,253 @@
+
+shader_type spatial;
+render_mode unshaded, shadows_disabled;
+void vertex() {
```
```suggestion

void vertex() {
```

AThousandShips replied:
```suggestion

void vertex() {
```

lyuma commented on scene/3d/chain_ik_3d.cpp:
```diff
@@ -0,0 +1,484 @@
+			p_property.usage = PROPERTY_USAGE_NONE;
+			force_hide = true;
+		}
+		if (force_hide || (split[2] == "end_bone" && !is_end_bone_extended(which) && split.size() > 3)) {
```
This will not hide `end_bone_name`
should `end_bone_name` should be inside `end_bone` (`end_bone/name`)?

lyuma commented on scene/3d/chain_ik_3d.cpp:
```diff
@@ -0,0 +1,484 @@
+		} else {
+			joints[p_joint].name = sk->get_bone_name(joints[p_joint].bone);
+		}
+	}
```
The setters behave differently depending on whether `get_skeleton()` is null.

I am worried there are cases which can desynchronize bone name/index

TokageItLab replied:
Same as below https://github.com/godotengine/godot/pull/110120#discussion_r2462824011

lyuma commented on scene/3d/chain_ik_3d.cpp:
```diff
@@ -0,0 +1,484 @@
+		props.push_back(PropertyInfo(Variant::STRING, path + "root_bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint));
+		props.push_back(PropertyInfo(Variant::INT, path + "root_bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR));
+		props.push_back(PropertyInfo(Variant::STRING, path + "end_bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint));
+		props.push_back(PropertyInfo(Variant::INT, path + "end_bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR));
```
Does this serialize both bone_name and bone (index)? I think we should pick one. We should only serialize bone_name.

TokageItLab replied:
Using both the bone name and index inherits the design of BoneAttachment. All other SkeletonModifier bone settings are serialized in the same way.

I remember discussing this with @fire in the past, and the main reasons for this design were to maintain compatibility with BoneAttachment and prevent glTF reimporting from breaking the indices (by making priority with the names).

It might be possible to remove this in the future, like in Godot5, but we should maintain consistency at least throughout Godot4. Therefore, this should not be changed in this PR.

If we want to change this, we should send a PR later that changes all other Nodes for Skeleton at the same time.

lyuma commented on scene/3d/chain_ik_3d.cpp:
```diff
@@ -0,0 +1,484 @@
+			set_root_bone(i, chain_settings[i]->root_bone.bone);
+		}
+		// Prior bone name.
+		if (!chain_settings[i]->end_bone.name.is_empty()) {
```
I think this should only run if `extend_end_bone` (since the property is undefined if not `extend_end_bone`)

TokageItLab replied:
Same as https://github.com/godotengine/godot/pull/110120#discussion_r2479446302

lyuma commented on scene/3d/chain_ik_3d.cpp:
```diff
@@ -0,0 +1,484 @@
+	ClassDB::bind_method(D_METHOD("get_joint_count", "index"), &ChainIK3D::get_joint_count);
+}
+
+void ChainIK3D::_validate_bone_names() {
```
I was surprised that this function does not directly call `_update_joints`, but it does not always need to.

It is confusing that `_skeleton_changed` in the subclass calls `_make_all_joints_dirty` which calls `_update_joints`.

This function calls `set_root_bone_name` which sometimes calls `set_root_bone` which sometimes calls `_update_joints`

The ordering and conditions for these update functions is quite complex for me to analyze. I would prefer if this flow is simpler and easier to understand. For example: when property is changed, if the skeleton is not null; or else the virtual functions when skeleton is changed (or _validate_bone_names)

TokageItLab replied:
The joint update logic uses nearly the same approach as SpringBoneSimulator3D.

The differences from SpringBoneSimulator3D are the use of a base class and the method for extending the struct array. However, the struct array-related methods adopted in this PR also reuse the design from the already-merged BoneConstraint3D.

In other words, it combines the designs of BoneConstraint3D and SpringBoneSimulator3D; this implementation is not being adopted for the first time in this PR.

What's slightly unique is that, unlike BoneConstraint3D, it requires an init process. Therefore, a virtual function is needed in the base class to call the init function for each struct array with different structs.

However, I couldn't find a better way to handle struct array properties with differing structures rather than current way.

This is due to conflicts between templates and inheritance. Considering compatibility with Godot's property registration, templates cannot be used extensively, making this approach necessary. Using Resources instead of Structs is another option, but it's probably not advisable from a performance and complexity standpoint.

To achieve cleaner code through other means would likely involve duplicating most of the init-related code across each class. However, this is essentially equivalent to splitting the ManyBoneIK class into five non-inherited classes, which I believe should not be done.

lyuma reviewed (commented):
I'm leaving some comments about property validation of ChainIK3D (the base class), since the logic and flow for bone-name <-> bone-index assignment, and updating of skeleton joints is overly complex and I fear there may be corner cases where the update is not done, or done multiple times.

Some examples of corner cases: if the bone name is assigned but not the bone index, or if the bone name and bone index are both assigned but set to different bones. And the same thing but before / after skeleton is assigned. And same thing but for joints but not root bone.

I wish that the flow could be made simpler. For example joint dirty flag + skeleton changed. Or go through a single update function.

I also would like if we can choose bone-name or bone-index, rather than supporting both, since allowing the user to set name or index (or both) adds even more complexity

AThousandShips commented on scene/3d/two_bone_ik_3d.cpp:
```diff
@@ -0,0 +1,782 @@
+		}
+		if (!valid) {
+			ERR_FAIL_EDMSG("End bone must be a child of middle bone.");
+		}
```
```suggestion
		ERR_FAIL_COND_EDMSG(!valid, "End bone must be a child of the middle bone.");
```

AThousandShips replied:
```suggestion
		ERR_FAIL_COND_EDMSG(!valid, "The end bone must be a child of the middle bone.");
```

TokageItLab commented:
We are making progress with focusing on 4.6, doing regular biweekly animation meetings on the contributor chat.

TokageItLab commented:
Rebased to the latest.

@lyuma I have responded to the code review comments, but fundamentally, the design of the ManyBoneIK class is based on other nodes (BoneConstraint3D and SpringBoneSimulator3D) already merged, as mentioned in https://github.com/godotengine/godot/pull/110120#discussion_r2470526241. Therefore, I assume there is not much we can specifically address here as it should be resolved at the same time as those other nodes.

I suppose the design of the internal struct mentioned in https://github.com/godotengine/godot/pull/110120#discussion_r2470526241 could be improved, but since the problem is how to deal the internal array, I believe it can definitely be sent as a future follow up without breaking compatibility.

lyuma commented on scene/3d/chain_ik_3d.cpp:
```diff
@@ -0,0 +1,484 @@
+		props.push_back(PropertyInfo(Variant::INT, path + "joint_count", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_ARRAY, "Joints," + path + "joints/,static,const"));
+		for (uint32_t j = 0; j < chain_settings[i]->joints.size(); j++) {
+			String joint_path = path + "joints/" + itos(j) + "/";
+			props.push_back(PropertyInfo(Variant::STRING, joint_path + "bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint, PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_READ_ONLY | PROPERTY_USAGE_STORAGE));
```
This should not be serialized, so remove STORAGE

lyuma reviewed (approved):
Discussed in animation meeting with @TokageItLab @SaracenOne @fire and @aaronfranke . I think we are good with merging this, and we can add blog post, do more testing and fix issues afterwards.

AThousandShips commented on doc/classes/CCDIK3D.xml:
```diff
@@ -0,0 +1,13 @@
+	</brief_description>
+	<description>
+		[CCDIK3D] is rotation based IK, enabling fast and effective tracking even with large joint rotations. It's especially suitable for chains with limitations, providing smoother and more stable target tracking compared to [FABRIK3D].
+		The result twist around forward vector will always be kept from previous pose.
```
```suggestion
		The resulting twist around the forward vector will always be kept from the previous pose.
```

AThousandShips replied:
```suggestion
		The resulting twist around the forward vector will always be kept from the previous pose.
```

TokageItLab commented:
The opinion is raised by other maintainers that the name ManyBoneIK3D is somewhat unclear.

Since the name SkeletonIK3D is unavailable until Godot 5, the following are the candidates proposed among the maintainers:

- IKGroup3D
- NewSkeletonIK3D
- SkeletonInverseKinematics3D
- InverseKinematics3D
- ModifierIK3D (current)

InverseKinematics3D might be acceptable as a superclass name, but considering Godot as a whole, there was an argument that it was too generic.

SkeletonInverseKinematics3D feels quite verbose, and considering the name FABRIK3D should not be SkeletonFABRIK3D, SkeletonFABRInverseKinematics3D or SkeletonForwardAndBackwardReachingInverseKinematics3D, I believe it should be avoided.

Considering that the words “Skeleton” and “Bone” are used as an identifier to avoid generality, the third identifier could be a "Modifier". Also, considering naming consistency with subclasses like "ChainIK3D", I tentatively preffer "ModifierIK3D".

Please suggest other name candidates if you have any.

ettiSurreal commented:
What about `IKModifier3D` since that seems to be a more consistent naming convention for skeleton modifiers that have the word "modifier" in their name. The only exception to that is `ModifierBoneTarget3D` which is the only one that has it at the start.

YahelBaram commented:
IKModifier3D
BonesIK3D

TokageItLab commented:
@ettiSurreal IKModifier was also a candidate, but considering that its subclasses are ChainIK and IterateIK, I thought consistency with them might suggest using “IK” as the suffix.

However, considering consistency with modifiers like LookAtModifier, “IKModifier” is fine. So it’s just a matter of whether to prioritize consistency with child or sibling classes.

Let's do quick poll on whether to use IKModifier3D or ModifierIK3D by reactions:

🚀 IKModifier3D
👀 ModifierIK3D

AThousandShips commented on scene/resources/3d/joint_limitation_3d.cpp:
```diff
@@ -0,0 +1,64 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "joint_limitation_3d.h"
```
```suggestion
#include "joint_limitation_3d.h"

```

AThousandShips commented on doc/classes/ChainIK3D.xml:
```diff
@@ -0,0 +1,144 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="ChainIK3D" inherits="ManyBoneIK3D" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+		A [SkeletonModifier3D] to apply inverse kinematics to bone chains containing arbitrary number of bones.
```
```suggestion
		A [SkeletonModifier3D] to apply inverse kinematics to bone chains containing an arbitrary number of bones.
```

AThousandShips commented on doc/classes/SplineIK3D.xml:
```diff
@@ -0,0 +1,86 @@
+	</methods>
+	<members>
+		<member name="setting_count" type="int" setter="set_setting_count" getter="get_setting_count" default="0">
+			The number of the settings.
```
```suggestion
			The number of settings.
```

AThousandShips commented on doc/classes/SplineIK3D.xml:
```diff
@@ -0,0 +1,86 @@
+			<description>
+				If [param size] is greater than [code]0[/code], the tilt is interpolated between [param size] of start bones from the start point of the [Curve3D] when they are apart.
+				If [param size] is equal [code]0[/code], the tilts between root bone head and the start point of the [Curve3D] are unified with a tilt of the start point of the [Curve3D].
+				If [param size] is less than [code]0[/code], the tilts between root bone and the start point of the [Curve3D] are [code]0.0[/code].
```
```suggestion
				If [param size] is less than [code]0[/code], the tilts between the root bone and the start point of the [Curve3D] are [code]0.0[/code].
```

AThousandShips commented on doc/classes/TwoBoneIK3D.xml:
```diff
@@ -0,0 +1,237 @@
+			<param index="1" name="bone_name" type="String" />
+			<description>
+				Sets the end bone name.
+				[b]Note:[/b] End bone must be a child of the middle bone.
```
```suggestion
				[b]Note:[/b] The end bone must be a child of the middle bone.
```

AThousandShips commented on doc/classes/TwoBoneIK3D.xml:
```diff
@@ -0,0 +1,237 @@
+			<param index="1" name="direction" type="int" enum="SkeletonModifier3D.SecondaryDirection" />
+			<description>
+				Sets the pole direction.
+				Pole is on the middle bone and will direct to the pole target.
```
```suggestion
				The pole is on the middle bone and will direct to the pole target.
```

AThousandShips commented on scene/3d/two_bone_ik_3d.cpp:
```diff
@@ -0,0 +1,775 @@
+	}
+
+	if (Math::is_equal_approx(Math::abs(kv.dot(fwd)), 1)) {
+		WARN_PRINT_ED("Setting: " + itos(p_index) + ": Pole direction and forward vectors are colinear. This is not advised as it may cause unwanted rotation.");
```
```suggestion
		WARN_PRINT_ED("Setting: " + itos(p_index) + ": Pole direction and forward vector are colinear. This is not advised as it may cause unwanted rotation.");
```

AThousandShips commented on editor/scene/3d/gizmos/chain_ik_3d_gizmo_plugin.cpp:
```diff
@@ -0,0 +1,253 @@
+	POSITION = PROJECTION_MATRIX * VIEW_MATRIX * MODEL_MATRIX * vec4(VERTEX.xyz, 1.0);
+	POSITION.z = mix(POSITION.z, POSITION.w, 0.998);
+}
+void fragment() {
```
```suggestion

void fragment() {
```

AThousandShips commented on doc/classes/JacobianIK3D.xml:
```diff
@@ -0,0 +1,13 @@
+	</brief_description>
+	<description>
+		[JacobianIK3D] calculates rotations for all joints simultaneously, producing natural and smooth movements. It is particularly suited for biological animations.
+		The result twist around forward vector will always be kept from previous pose.
```
```suggestion
		The resulting twist around the forward vector will always be kept from the previous pose.
```

AThousandShips commented on scene/3d/two_bone_ik_3d.h:
```diff
@@ -0,0 +1,310 @@
+
+		NodePath pole_node;
+		SecondaryDirection pole_direction = SECONDARY_DIRECTION_NONE; // Sort to pole target plane.
+		Vector3 pole_direction_vector = Vector3(0, 0, 0); // Custom vector.
```
```suggestion
		Vector3 pole_direction_vector; // Custom vector.
```

AThousandShips commented on scene/3d/spline_ik_3d.cpp:
```diff
@@ -0,0 +1,427 @@
+	PackedStringArray warnings = SkeletonModifier3D::get_configuration_warnings();
+	for (uint32_t i = 0; i < sp_settings.size(); i++) {
+		if (sp_settings[i]->path_3d.is_empty()) {
+			warnings.push_back(RTR("Detect settings with no Path3D set! SplineIK3D must be set Path3D to work."));
```
```suggestion
			warnings.push_back(RTR("Detecting settings with no Path3D set! SplineIK3D must have a Path3D to work."));
```

AThousandShips commented on scene/3d/two_bone_ik_3d.cpp:
```diff
@@ -0,0 +1,775 @@
+	}
+	for (uint32_t i = 0; i < tb_settings.size(); i++) {
+		if (tb_settings[i]->target_node.is_empty()) {
+			warnings.push_back(RTR("Detect settings with no pole target set! TwoBoneIK3D must be set pole target to work."));
```
```suggestion
			warnings.push_back(RTR("Detecting settings with no pole target set! TwoBoneIK3D must have a pole target to work."));
```

AThousandShips commented on doc/classes/SplineIK3D.xml:
```diff
@@ -0,0 +1,86 @@
+			<return type="bool" />
+			<param index="0" name="index" type="int" />
+			<description>
+				Returns the tilt property of [Curve3D] takes effect to the bone twist.
```
```suggestion
				Returns if the tilt property of the [Curve3D] affects the bone twist.
```

AThousandShips commented on doc/classes/SplineIK3D.xml:
```diff
@@ -0,0 +1,86 @@
+		A [SkeletonModifier3D] for aligning bones on [Path3D]. The smoothness of the fitting depends on the [member Curve3D.bake_interval].
+		If you want [Path3D] to attach a specific bone, it is recommended to place [ModifierBoneTarget3D] before [SplineIK3D] in the [SkeletonModifier3D] list (children of the [Skeleton3D]), and then place [Path3D] as [ModifierBoneTarget3D]'s child.
+		Bone twist is determined based on the [method Curve3D.get_point_tilt].
+		If the root bone joint and the start point of [Curve3D] are apart, it assumes that there is a linear line segment between them. It means that the vector pointing toward the start point of the [Curve3D] takes precedence over the shortest intersection point along the [Curve3D].
```
```suggestion
		If the root bone joint and the start point of the [Curve3D] are separated, it assumes that there is a linear line segment between them. This means that the vector pointing toward the start point of the [Curve3D] takes precedence over the shortest intersection point along the [Curve3D].
```

AThousandShips commented on doc/classes/SplineIK3D.xml:
```diff
@@ -0,0 +1,86 @@
+			<return type="int" />
+			<param index="0" name="index" type="int" />
+			<description>
+				Returns the tilt interpolation method between end bone and the end point of the [Curve3D] when they are apart. See also [method set_tilt_fade_out]
```
```suggestion
				Returns the tilt interpolation method used between the end bone and the end point of the [Curve3D] when they are apart. See also [method set_tilt_fade_out].
```

AThousandShips commented on doc/classes/JacobianIK3D.xml:
```diff
@@ -0,0 +1,13 @@
+		Pseudo inverse Jacobian matrix based inverse kinematics solver.
+	</brief_description>
+	<description>
+		[JacobianIK3D] calculates rotations for all joints simultaneously, producing natural and smooth movements. It is particularly suited for biological animations.
```
```suggestion
		[JacobianIK3D] calculates rotations for all joints simultaneously, producing natural and smooth movement. It is particularly suited for biological animations.
```

AThousandShips commented on doc/classes/CCDIK3D.xml:
```diff
@@ -0,0 +1,13 @@
+	<description>
+		[CCDIK3D] is rotation based IK, enabling fast and effective tracking even with large joint rotations. It's especially suitable for chains with limitations, providing smoother and more stable target tracking compared to [FABRIK3D].
+		The result twist around forward vector will always be [code]0[/code].
+		[b]Note:[/b] When the target is close to the root, it can cause unnatural movements, including joint flips and oscillations.
```
```suggestion
		[b]Note:[/b] When the target is close to the root, it can cause unnatural movement, including joint flips and oscillations.
```
"movement" is uncountable in this case

AThousandShips commented on scene/3d/two_bone_ik_3d.h:
```diff
@@ -0,0 +1,310 @@
+		}
+
+		// Make rotation as bone pose from chain coordinates.
+		void cache_current_joint_rotations(Skeleton3D *p_skeleton, Vector3 p_pole_destination) {
```
```suggestion
		void cache_current_joint_rotations(Skeleton3D *p_skeleton, const Vector3 &p_pole_destination) {
```

AThousandShips commented on doc/classes/ChainIK3D.xml:
```diff
@@ -0,0 +1,144 @@
+			<return type="int" enum="SkeletonModifier3D.BoneDirection" />
+			<param index="0" name="index" type="int" />
+			<description>
+				Returns the end bone's tail direction of the bone chain when [method is_end_bone_extended] is [code]true[/code].
```
```suggestion
				Returns the tail direction of the end bone of the bone chain when [method is_end_bone_extended] is [code]true[/code].
```
Or
```suggestion
				Returns the end bone tail direction of the bone chain when [method is_end_bone_extended] is [code]true[/code].
```

rendlp commented:
Thank god for this. LookUpModifier3D ain't it.

ettiSurreal commented:
@TokageItLab Also, if SkeletonIK3D does get removed in the future, I think we should just keep "IKModifier3D" as this modifier's name. I think it works better than "SkeletonIK3D", is more consistent with other modifiers, and by that point users will be used to the name of the new modifier.

TokageItLab commented:
> So I gave this a test, but the results seem far less stable than the existing SkeletonIK3D. I think part of the problem might be that there's no way to to make the IK target follow the rotation of the target, like the "Override Tip Basis" setting on SkeletonIK3D.

@jitspoe As I mentioned in the description, we consider this to be the role of BoneConstraint rather than IK, so it is supplemented by https://github.com/godotengine/godot/pull/110336.

Another possible cause is when the change of the bone pose position or scale from the bone rest is included in the chain, but that needs to be supplemented by https://github.com/godotengine/godot/pull/111815.

YahelBaram commented:
Hi

I'm currently using Godot 4.6 dev 4, and I'm doing some tests for the new IKs system. While I'm pretty happy with the results when it comes to bones positioning, I don't see anything related to rotation (For example, rotating the hand of a player model in a way the entire arm will react to the changes of the hand) . I wanted to know if I'm missing something or if it's not in the scope of this PR in the first place.

Thanks ahead!

TokageItLab commented:
@YahelBaram That is considered more the role of BoneConstraint than IK, and this was agreed upon during the animation meeting.

For example, when the twist axis is +Y, the propagation of hand rotation can be fulfilled by transferring the +Y rotation of the hand to the +Y axis rotation of the LowerArm/UpperArm using BoneConstraint(Copy/ConvertTransformModifier3D). See also https://github.com/godotengine/godot/pull/100984, https://github.com/godotengine/godot/pull/110336 and https://github.com/godotengine/godot/pull/111367.

TokageItLab commented:
@x0r-b0t It seems to be an issue with the dynamic bone axis direction calculation by [mutable_bone_axis](https://github.com/godotengine/godot/pull/111815). I'll take a look.
